### DISCUSSION TITLE
Nightly builds

### DISCUSSION THREAD

**Discussion by MarekPikula** [SpinalHDL/SpinalHDL#793] [Ideas]
@Dolu1990 I'm wondering whether it would be possible to publish nightly builds of SpinalHDL on Maven. I'm currently using a locally published git version because I'm affected by a bug fixed with [this commit](https://github.com/SpinalHDL/SpinalHDL/commit/b1543a34275a5103f884967f12addf74db52156d). It would be very nice if there was a periodic GitHub Workflow publishing current dev as a nightly snapshot, so it can be used without local publishing.

Local publishing is not that big of a deal when working locally, but the project I'm working on has a CI pipeline, which would need to do the same step. With multiple runners, it gets cumbersome and time-consuming, as each runner needs to build and cache the locally published version before running the actual pipeline.

**Comment by MarekPikula** [comment]
It would also open a way for an automated release deployment via GitHub Workflows.

**Reply by MarekPikula** [reply]
I just noticed that there is already Travis script for deployment, so I guess it would be relatively easy to make it build directly on dev.

**Reply by Dolu1990** [reply]
hmm, in some ways, i would feel bad to spam Maven repositories with big fat jar, 365 times per year, i don't know if maven is realy made for such usage.

So far, i'm kind of never using publish local, but always integrating SpinalHDL from source in my projects, as it also allow to easily modify it.

Maybe we could just release SpinalHDL minor version at a faster rate ?

**Reply by MarekPikula** [reply]
I'm currently using Mill for build management, and there seems to be no easy way to “import” an external sbt project as a dependency (as mentioned, e.g., [here](https://github.com/SpinalHDL/SpinalCrypto/issues/10)). I guess I could go back to sbt, but that would be somewhat painful (sbt seems to be slower than mill in some areas, especially when ran directly as a command).

Is there no alternative to Maven repo? How are other big libraries handling nightly builds (I guess some need to have this option, right?)

Pushing patch versions faster seems to be an option, but I think that it would quickly become messy and tedious to push a new version after near every bugfix.

**Reply by MarekPikula** [reply]
Or maybe do you know if there is some sort of local Maven repo alternative? Something to set up, to have it as artefact repo for CI.

**Reply by Dolu1990** [reply]
> I'm currently using Mill for build management, and there seems to be no easy way to “import” an external sbt project as a dependency (as mentioned, e.g., https://github.com/SpinalHDL/SpinalCrypto/issues/10)

Ahhh make sense

> Is there no alternative to Maven repo? 

So, personnaly, i publish to sonatype, and things endup on maven, but all of this is mostly blackmagic for me XD

> How are other big libraries handling nightly builds

I have no idea.

> Pushing patch versions faster seems to be an option, but I think that it would quickly become messy and tedious to push a new version after near every bugfix.

Yes ^^
I mean, maybe not more than once each two weeks ?
What i did for SpinalHDL 1.7.0 is to release 1.7.0a 1.7.0b for hot fixes, this would keep things light.

> Or maybe do you know if there is some sort of local Maven repo alternative? Something to set up, to have it as artefact repo for CI.

I guess that's probably possible, but that's realy out of my knowledge

**Reply by MarekPikula** [reply]
I researched a bit, and it seems that the standard practice is to publish SNAPSHOT releases. It seems that Sonatype has an option for that: https://s01.oss.sonatype.org/content/repositories/snapshots/, including [clean-up policy management](https://help.sonatype.com/repomanager3/nexus-repository-administration/tasks/maven-snapshot-tasks). It seems that it doesn't go directly to Maven, but rather to a separate, snapshot repository.

**Reply by Dolu1990** [reply]
Ahhh interresting i wasn't aware of that feature ^^
Will have to try that.

Also, i just released 1.7.2, should be on maven in the next hour.

**Reply by MarekPikula** [reply]
Nice :+1:

**Reply by MarekPikula** [reply]
Mind that you didn't push a release on GitHub :wink:

**Reply by Dolu1990** [reply]
[Done](https://github.com/SpinalHDL/SpinalHDL/releases/tag/v1.7.2) done ^^

**Comment by Readon** [comment]
Would you please release the URL of the repository for SNAPSHOT?
I have checked [OSS repos](https://oss.sonatype.org/content/repositories/snapshots/com/github/spinalhdl/) and the link above and could not find the SNAPSHOT released.

**Reply by Dolu1990** [reply]
Ahh i think it is because never any snapshot was released online.
I mean, i never did such release.

Wanna me to try ?

**Reply by Readon** [reply]
it doesn't matter, I can use publishLocal still.
